### PR TITLE
Qprofiler may2019

### PR DESCRIPTION
**Pull request by ChristinaXu2017**
tidy up code; update xsd file; fixed two bugs during update map for tags. now both QNAM and TAG use same method to update map which method allows to set the map limit size.

### CODE REVIEW

File: qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java [hunk 1]
```diff
@@ -236,9 +239,25 @@ public static Element outputBins( Element parent, String name, Map<Integer, Atom
...
-    	return cateEle;
-    	
+    	return cateEle;   	
     }
+	
+	public static <T> void outputTallyGroupWithSize(Element parent, String name, Map<T, AtomicLong> tallys, int sizeLimits) {
+		String comment = null; 
+		boolean hasPercent =  true;
+		String tallyCount = null; 
+		if( tallys.size() > sizeLimits) {
```
Comment (holmeso) [inline qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java]: what if `tallys` is null? Should always check inputs of static methods

File: qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java [hunk 2]
```diff
@@ -299,16 +318,34 @@ public static Element createReadGroupNode( Element parent, String rgid) {
...
+	/**
+	 *  
+	 * @param map: Map<String, AtomicLong>
+	 * @param key: key string of map
+	 * @return true if it is a new key added
+	 */
+	public static boolean updateMapWithLimit(Map<String, AtomicLong> map , String key, int limitSize) {
+		boolean isNew = false; 
+		String key1 = key;	 	
+		if(!map.containsKey(key1)) {
```
Comment (holmeso) [inline qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java]: `map` could be null

File: qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java [hunk 3]
```diff
@@ -299,16 +318,34 @@ public static Element createReadGroupNode( Element parent, String rgid) {
...
      * @param str2 will covert the first letter to upper case
      * @return a joined string
      */
-    public static String join(String str1, String str2) {
-    	
-    	return str1 + str2.substring(0,1).toUpperCase() + str2.substring(1);
-    }
-        
-        
+//    public static String join(String str1, String str2) {
```
Comment (holmeso) [inline qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java]: delete

File: qprofiler2/src/org/qcmg/qprofiler2/bam/TagSummaryReport2.java [hunk 4]
```diff
@@ -113,29 +106,21 @@ public void toXml(Element parent){
 		
 		// additional tags includes RG
 		for (Entry<String,  ConcurrentSkipListMap<String, AtomicLong>> entry : additionalTags.entrySet())	
+			//XmlUtils.outputTallyGroupWithSize(parent, entry.getKey(), entry.getValue(), ADDI_TAG_MAP_LIMIT);
```
[the remark this answers is not in the document]
Comment (ChristinaXu2017) [reply qprofiler2/src/org/qcmg/qprofiler2/bam/TagSummaryReport2.java]: done

File: qprofiler2/src/org/qcmg/qprofiler2/bam/TagSummaryReport2.java [hunk 5]
```diff
@@ -113,29 +106,21 @@ public void toXml(Element parent){
...
 			outputTag(parent, entry.getKey(),  entry.getValue());
-		
-		// additional tagsInt
-		for (Entry<String,  QCMGAtomicLongArray> entry : additionalIntegerTags.entrySet())
-			outputTag(parent,  entry.getKey(),  entry.getValue().toMap());
-		
+
```
[the remark this answers is not in the document]
Comment (ChristinaXu2017) [reply qprofiler2/src/org/qcmg/qprofiler2/bam/TagSummaryReport2.java]: done

File: qprofiler2/src/org/qcmg/qprofiler2/summarise/ReadIDSummary.java [hunk 6]
```diff
@@ -325,13 +323,13 @@ public void toXml(Element ele){
...
 	}
 	
-	private void outputTallyGroup(Element parent, String name, Map<String, AtomicLong> tallys) {
-		Element e1 = XmlUtils.outputTallyGroup( parent, name, tallys , false );
-		if(e1 != null ) {
-			String v1 = (tallys.size() < TALLY_SIZE )? tallys.size()+"" : TALLY_SIZE + "+";
-			e1.setAttribute("no", v1);
-		}
-	}
+//	private void outputTallyGroup(Element parent, String name, Map<String, AtomicLong> tallys) {
```
[the remark this answers is not in the document]
Comment (ChristinaXu2017) [reply qprofiler2/src/org/qcmg/qprofiler2/summarise/ReadIDSummary.java]: done

File: qprofiler2/src/org/qcmg/qprofiler2/summarise/ReadIDSummary.java [hunk 7]
```diff
@@ -340,17 +338,17 @@ private void outputTallyGroup(Element parent, String name, Map<String, AtomicLon
...
-		if(!map.containsKey(key1)) { 
-			if(map.size() >= TALLY_SIZE ) {key1 = OTHER;}
-			isNew = true; 
-		} 
-		
-		map.computeIfAbsent(key1, k -> new AtomicLong()).incrementAndGet();
-		 
-		return isNew; 
-	}	
+//	private  boolean updateMap(ConcurrentMap<String, AtomicLong> map , String key) {
```
[the remark this answers is not in the document]
Comment (ChristinaXu2017) [reply qprofiler2/src/org/qcmg/qprofiler2/summarise/ReadIDSummary.java]: done

File: qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java [hunk 8]
```diff
@@ -30,6 +32,7 @@
 	public static final String Scount = "count";
 	public static final String Spercent = "percent";
 	public static final String Stally = "tally";
+	public static final String StallyCount = "tallyCount";
```
[the remark this answers is not in the document]
Comment (ChristinaXu2017) [reply qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java]: done

File: qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java (hunk 2 above)
Comment (ChristinaXu2017) [reply qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java]: done

File: qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java (hunk 3 above)
Comment (ChristinaXu2017) [reply qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java]: done

File: qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java (hunk 1 above)
Comment (ChristinaXu2017) [reply qprofiler2/src/org/qcmg/qprofiler2/util/XmlUtils.java]: done

**Comment by ChristinaXu2017** [comment]
enable index option, tidy up, delete debug code

**Pull request merged**